diffusion/models/models.py

```diff
@@ -580,6 +581,283 @@ def stable_diffusion_xl(
+    for block in unet.down_blocks:
+        if hasattr(block, 'attentions'):
+            for attention in block.attentions:
+                attention._fsdp_wrap = True
+        if hasattr(block, 'resnets'):
+            for resnet in block.resnets:
+                resnet._fsdp_wrap = True
+
+    # Make the noise schedulers
+    train_scheduler_params: Dict[str, Any] = {
+        'num_train_timesteps': 1000,
+        'beta_start': 0.00085,
+        'beta_end': 0.012,
+        'beta_schedule': 'scaled_linear',
+        'variance_type': 'fixed_small',
+        'clip_sample': False,
+        'prediction_type': prediction_type,
+        'sample_max_value': 1.0,
+        'timestep_spacing': 'leading',
+        'steps_offset': 1,
+        'rescale_betas_zero_snr': False,
+    }
+    if train_noise_scheduler_params is not None:
+        train_scheduler_params.update(train_noise_scheduler_params)
+    noise_scheduler = DDPMScheduler(**train_scheduler_params)
+
+    inference_scheduler_params: Dict[str, Any] = {
+        'num_train_timesteps': 1000,
+        'beta_start': 0.00085,
+        'beta_end': 0.012,
+        'beta_schedule': 'scaled_linear',
+        'trained_betas': None,
+        'prediction_type': prediction_type,
+        'interpolation_type': 'linear',
+        'use_karras_sigmas': False,
+        'timestep_spacing': 'leading',
+        'steps_offset': 1,
+        'rescale_betas_zero_snr': False,
+    }
+
+    if inference_noise_scheduler_params is not None:
+        inference_scheduler_params.update(inference_noise_scheduler_params)
+    inference_noise_scheduler = EulerDiscreteScheduler(**inference_scheduler_params)
+
+    # Shift noise scheduler to correct for resolution changes
+    noise_scheduler = shift_noise_schedule(noise_scheduler,
+                                           base_dim=32,
+                                           shift_dim=scheduler_shift_resolution // downsample_factor)
+    inference_noise_scheduler = shift_noise_schedule(inference_noise_scheduler,
+                                                     base_dim=32,
+                                                     shift_dim=scheduler_shift_resolution // downsample_factor)
+
+    # Optionally load the tokenizers and text encoders
+    t5_tokenizer, t5_encoder, clip_tokenizer, clip_encoder = None, None, None, None
+    if include_text_encoders:
+        t5_tokenizer = AutoTokenizer.from_pretrained('google/t5-v1_1-xxl', cache_dir=cache_dir, local_files_only=True)
+        clip_tokenizer = AutoTokenizer.from_pretrained('stabilityai/stable-diffusion-xl-base-1.0',
+                                                       subfolder='tokenizer',
+                                                       cache_dir=cache_dir,
+                                                       local_files_only=False)
+        t5_encoder = AutoModel.from_pretrained('google/t5-v1_1-xxl',
+                                               torch_dtype=torch.float16,
```

Comment:
hmm should be bfloat16? or read the dtype from an argument?

Reply:
Changed it to be an argument, but left the default as-is for inference.